### PR TITLE
[SPARK-14603] [SQL] Verification of Metadata Operations by Session Catalog

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

Since we cannot really trust if the underlying external catalog can throw exceptions when there is an invalid metadata operation, let's do it in SessionCatalog.
- [X] The first step is to unify the error messages issued in Hive-specific Session Catalog and general Session Catalog. 
- [X] The second step is to verify the inputs of metadata operations for partitioning-related operations. This is moved to a separate PR: https://github.com/apache/spark/pull/12801
- [X] The third step is to add database existence verification in `SessionCatalog`
- [X] The fourth step is to add table existence verification in `SessionCatalog`
- [X] The fifth step is to add function existence verification in `SessionCatalog`
#### How was this patch tested?

Add test cases and verify the error messages we issued

### CODE REVIEW

**Comment by gatorsmile** [comment]
Partition-related operations are complex. 
- Drop Partitions: It allows partial matches, however, we need to ensure the orders are the same. That means, the specified to-be-dropped spec should be consistent with the defined partitioning spec. 
- Add Partitions: We need to ensure the newly added columns are from the existing schema. We also need to ensure it does not miss the leading partitioning columns. 
- Rename Partitions: The old spec should be part of the existing partitioning spec. The new spec should be part of table schema.
... Still trying to explore all the cases in Apache Hive. @yhuai @andrewor14

**Comment by gatorsmile** [comment]
```
hive> ALTER TABLE extTable_with_partitions ADD PARTITION (ds='2008-04-08', hr='12', key='1')
    > ;
FAILED: ValidationFailureSemanticException Partition spec {ds=2008-04-08, hr=12, key=1} contains non-partition columns
```
Another error we did not catch in the current master build.

**Comment by gatorsmile** [comment]
The test cases are wrong, based on the Hive definition. 
- `Rename Partitions` only allows users to change the partition values, instead of partition column names.
- `Add Partitions` only allows users to add new partitions with the exactly same partition columns defined in Table. 
- `Alter Partition Set Location` also need to ensure users to use the same partitioning definition. 

@andrewor14 @yhuai Do you want me to submit a separate PR to fix all the issues in the implementation of Alter Partition? In the `InMemoryCatalog`, we ignore all the above issues.

**Comment by yhuai** [comment]
@gatorsmile Yea, you are right. Those calls should use already defined partition column. Otherwise, we need to throw errors. If fixing the tests will not cause too many lines of changes, can you do it in this PR? Or, we can also do it in a separate PR.

**Comment by gatorsmile** [comment]
Thank you! The work requires more and more code changes. Let me split the work into two parts. One is for handling partitioning-related PR; another is for handling the remaining parts.

**Comment by gatorsmile** [comment]
@yhuai @andrewor14 The latest commit moves DB existence checking from external catalog to session catalog. I am not sure if this is the right move. I need to hear your inputs before continuing the work to move the other similar checking. 

So far, Hive external catalog has very limited input checking, like DB checking. It relies on the Hive. So far, the test cases in `HiveExternalCatalogSuite` does not check if the message contents are expected. (I know, it is hard if `HiveExternalCatalogSuite` and `InMemoryCatalogSuite` share the same test cases) Thus, the current way could miss capturing the bugs. This is one of the potential issues we are resolving now.

Assuming users might extend and use the other external catalog, we definitely should do the checking in `SessionCatalog`. It will improve the quality since the implementation of these external catalog is not predictable. 

To avoid duplicate checking, we only do the checking in `SessionCatalog`. **This assumes `SessionCatalog` is the ONE AND ONLY component that uses `HiveExternalCatalog` and `InMemoryCatalog`.** I am not sure if this assumption is valid and we will keep them valid. : )

Let me know if you have any opinion. Thanks!

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
ping @yhuai Could you check if my question? I need to get the answer before continuing the work. https://github.com/apache/spark/pull/12385#issuecomment-210968627 

Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala [hunk 1]
```diff
@@ -107,18 +105,16 @@ class InMemoryCatalog extends ExternalCatalog {
...
     }
   }
 
   override def alterDatabase(dbDefinition: CatalogDatabase): Unit = synchronized {
-    requireDbExists(dbDefinition.name)
     catalog(dbDefinition.name).db = dbDefinition
   }
 
   override def getDatabase(db: String): CatalogDatabase = synchronized {
-    requireDbExists(db)
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: why remove these?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 2]
```diff
@@ -243,10 +257,10 @@ class SessionCatalog(
     if (name.database.isDefined || !tempTables.contains(table)) {
       // When ignoreIfNotExists is false, no exception is issued when the table does not exist.
       // Instead, log it as an error message.
-      if (externalCatalog.tableExists(db, table)) {
+      if (tableExists(name)) {
         externalCatalog.dropTable(db, table, ignoreIfNotExists = true)
       } else if (!ignoreIfNotExists) {
-        logError(s"Table or View '${name.quotedString}' does not exist")
+        logError(new NoSuchTableException(db = db, table = table).getMessage)
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: I fixed this in my patch #12853, but we should throw exception here.

**Comment by andrewor14** [comment]
> To avoid duplicate checking, we only do the checking in SessionCatalog. This assumes SessionCatalog is the ONE AND ONLY component that uses HiveExternalCatalog and InMemoryCatalog.

Hey @gatorsmile while this is likely I don't think we should assume that. Right now the boundaries are pretty clear and the two classes can be tested separately. I don't think it's a good idea to add an implicit dependency there. Could you revert those changes?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.catalog.ExternalCatalog.TablePartitionSpec
+
+/**
+ * Thrown by a catalog when an item already exists. The analyzer will rethrow the exception
+ * as an [[org.apache.spark.sql.AnalysisException]] with the correct position information.
+ */
+class AlreadyExistDatabaseException(db: String)
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala]: I like this idea. I would call this `DatabaseAlreadyExistsException` though (and also rename the other ones).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  extends AnalysisException(s"Database '$db' already exists")
+
+class AlreadyExistTableException(db: String, table: String)
+  extends AnalysisException(s"Table or View '$table' already exists in database '$db'")
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala]: small nit: can you make it lower case view (I had to fix this in my other patch too)

**Comment by andrewor14** [comment]
> To avoid duplicate checking, we only do the checking in SessionCatalog. This assumes SessionCatalog is the ONE AND ONLY component that uses HiveExternalCatalog and InMemoryCatalog.

Hey @gatorsmile while this is likely I don't think we should assume that. Right now the boundaries are pretty clear and the two classes can be tested separately. I don't think it's a good idea to add an implicit dependency there. Could you revert those changes?

**Comment by gatorsmile** [comment]
@andrewor14 Sure, will revert the changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala]: Will fix it later. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 5]
```diff
@@ -398,6 +432,14 @@ class SessionCatalog(
     }
   }
 
+ /**
+  * Return whether a temp table with the specified name exists.
+  */
+  def tempTableExists(name: TableIdentifier): Boolean = {
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: delete: not used anywhere

File: python/pyspark/sql/utils.py [hunk 6]
```diff
@@ -61,6 +61,8 @@ def deco(*a, **kw):
                                              e.java_exception.getStackTrace()))
             if s.startswith('org.apache.spark.sql.AnalysisException: '):
                 raise AnalysisException(s.split(': ', 1)[1], stackTrace)
+            if s.startswith('org.apache.spark.sql.catalyst.analysis.NoSuchTableException: '):
```
Comment (andrewor14) [inline python/pyspark/sql/utils.py]: we catch only this specific one in python? Why not all the other ones too?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 7]
```diff
@@ -111,6 +111,25 @@ class SessionCatalog(
     fs.makeQualified(hadoopPath)
   }
 
+  protected[this] def requireDbExists(db: String): Unit = {
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: why make them protected again? These should just be private right?

**Comment by andrewor14** [comment]
This patch by itself looks OK to me so I'm going to merge it into master 2.0 first. Feel free to submit follow-up patches to fix the rest of the comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 5 above)
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: I'm deleting this for oyu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Yeah, we can change it, unless HiveSessionCatalog uses it.

File: python/pyspark/sql/utils.py (hunk 6 above)
Comment (gatorsmile) [inline python/pyspark/sql/utils.py]: It caused a test case failure. 

Yeah, we can add all of them into this file.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Thank you!

**Comment by gatorsmile** [comment]
Thank you, will do the follow-up PR.

File: python/pyspark/sql/utils.py (hunk 6 above)
Comment (andrewor14) [inline python/pyspark/sql/utils.py]: which test?

Comment (andrewor14) [inline python/pyspark/sql/utils.py]: I think we should either add all or nothing. I kind of prefer nothing actually because it's less code to maintain.

Comment (gatorsmile) [inline python/pyspark/sql/utils.py]: The failed test case is https://github.com/apache/spark/blob/29314379729de4082bd2297c9e5289e3e4a0115e/python/pyspark/sql/tests.py#L1564-L1567

I checked the history. The reason why we captured the exceptions is for issuing a better error message. See the original PR: https://github.com/apache/spark/pull/7135

I will try to add all the exception types. Thanks!

**Pull request closed without merging**